### PR TITLE
Adding a migration test for no interest zoning scenario

### PR DESCRIPTION
**Pull request by MatthewSandfordImprobable**
Using the existing migration test with some add-ons to test migration in a zero interest zoning LB strategy. 

Tests has an actor that runs between zones. The test checks that the actor makes it to  the destination point.
The test also checks the speed of the actor is never higher than a threshold.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Private/TestMaps/Spatial2WorkerFullInterestMigrationMap.cpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	// Add the tests
+	AddActorToLevel<ASpatialTestCharacterMigration>(CurrentLevel, FTransform::Identity);
+
+	FTransform PlayerTransform;
+
+	PlayerTransform.SetTranslation(FVector(-500.0f, -500.0f, 50.0f));
+	AddActorToLevel<APlayerStart>(CurrentLevel, PlayerTransform);
+
+	PlayerTransform.SetTranslation(FVector(-500.0f, 500.0f, 50.0f));
+	AddActorToLevel<APlayerStart>(CurrentLevel, PlayerTransform);
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/Private/TestMaps/Spatial2WorkerFullInterestMigrationMap.cpp:29]: feels like this could be refactored so the two maps descend from something common

Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/Private/TestMaps/Spatial2WorkerFullInterestMigrationMap.cpp:29]: but it could be more work than it's worth. though you could combine the source files so it doesn't sprawl that many places.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp [hunk 2]
```diff
@@ -38,19 +38,29 @@ ASpatialTestCharacterMigration::ASpatialTestCharacterMigration()
 	Author = "Victoria";
 	Description = TEXT("Test Character Migration");
 	TimeLimit = 300;
+
+	Destination = FVector(200.0f, 0.0f, 40.0f);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp:42]: I'd rather these were set from `UAbstractLBStrategy::GetWorkerEntityPosition`

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp:42]: Or at least some validation that Destination and Origin are on different servers and handover is taking place. You could monitor authgained/lost calls, or use `WhoShouldHaveAuthority` perhaps

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp [hunk 3]
```diff
@@ -76,89 +86,84 @@ void ASpatialTestCharacterMigration::PrepareTest()
...
 	});
 
-	// Repeatedly move character forwards and backwards over the worker boundary and adding actors every time
+	AddStepFromDefinition(WaitForStationaryActorStepDefinition, FWorkerDefinition::AllClients);
 	for (int i = 0; i < 5; i++)
 	{
-		if (i < 1)
+		if (i == 0)
 		{
 			AddStepFromDefinition(AddActorStepDefinition, FWorkerDefinition::AllServers);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp:157]: Just do this before the loop?

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp:157]: Oh - it matches the if statement below - fine

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMovement/CharacterMovementTestGameMode.h [hunk 4]
```diff
@@ -13,4 +13,13 @@ UCLASS()
...
 	GENERATED_UCLASS_BODY()
+
+	virtual void BeginPlay() override;
+
+	// This is overriden so that we can ensure the correct player starts are used in tests.
+	AActor* FindPlayerStart_Implementation(AController* Player, const FString& IncomingName) override;
+
+private:
+	int32 SpawnRequests = 0;
+	TArray<APlayerStart*> PlayerStarts;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMovement/CharacterMovementTestGameMode.h:24]: UPROPERTY

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/TestActors/TestMovementCharacter.cpp [hunk 5]
```diff
@@ -41,6 +44,55 @@ ATestMovementCharacter::ATestMovementCharacter()
...
+{
+	Super::BeginPlay();
+	PreviousLocation = GetActorLocation();
+}
+
+void ATestMovementCharacter::Tick(float DeltaTime)
+{
+	float Speed = FVector(GetActorLocation() - PreviousLocation).Size() / DeltaTime;
+
+	if (SpeedWindowIndex > SpeedWindowSize - 1)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/TestActors/TestMovementCharacter.cpp:64]:
```suggestion
	if (SpeedWindowIndex >= SpeedWindowSize )
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/TestActors/TestMovementCharacter.cpp [hunk 6]
```diff
@@ -41,6 +44,55 @@ ATestMovementCharacter::ATestMovementCharacter()
...
+	{
+		SpeedWindowIndex = 0;
+	}
+
+	SpeedWindow[SpeedWindowIndex] = Speed;
+	SpeedWindowIndex++;
+
+	PreviousLocation = GetActorLocation();
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/TestActors/TestMovementCharacter.cpp:72]: nit - can cache GetActorLocation() above and just call it once

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/TestActors/TestMovementCharacter.cpp [hunk 7]
```diff
@@ -41,6 +44,55 @@ ATestMovementCharacter::ATestMovementCharacter()
...
+}
+
+float ATestMovementCharacter::GetPeakSpeedInWindow() const
+{
+	float Max = 0.0f;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/TestActors/TestMovementCharacter.cpp:77]: Did you measure how many slots in the window are filled between checks?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp [hunk 8]
```diff
@@ -25,11 +25,11 @@ float GetTargetDistanceOnLine(const FVector& From, const FVector& Target, const
 } // namespace
 
 /**
- * This test moves a character backward and forward repeatedly between two workers, adding actors. Based on the SpatialTestCharacterMovement
- * test. This test requires the CharacterMovementTestGameMode, trying to run this test on a different game mode will fail.
- *
- * The test includes two servers and one client worker. The client worker begins with a PlayerController and a TestCharacterMovement
- *
+ * This test moves a character backward and forward repeatedly between two workers ensuring migration occurs.
+ * PlayerController owned actors are spawned on both worker to ensure owned actors do not hinder migration.
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp:29]:
```suggestion
 * PlayerController owned actors are spawned on both workers to ensure owned actors do not hinder migration.
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp [hunk 9]
```diff
@@ -25,11 +25,11 @@ float GetTargetDistanceOnLine(const FVector& From, const FVector& Target, const
...
+ * Based on the SpatialTestCharacterMovement test. This test requires the CharacterMovementTestGameMode; trying to run this test on a
+ * different game mode will fail. The test includes two servers and one client worker. Each client worker begins with a PlayerController and
+ * a TestCharacterMovement.
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp:32]:
```suggestion
 * a TestMovementCharacter.
```
?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp [hunk 10]
```diff
@@ -38,19 +38,29 @@ ASpatialTestCharacterMigration::ASpatialTestCharacterMigration()
...
-		bCharacterReachedDestination = false;
-		bCharacterReachedOrigin = false;
+	// Wait for actor to be stationary
+	FSpatialFunctionalTestStepDefinition WaitForStationaryActorStepDefinition(/*bIsNativeDefinition*/ true);
+	WaitForStationaryActorStepDefinition.StepName = TEXT("WaitForStationaryActorStepDefinition");
+	WaitForStationaryActorStepDefinition.TimeLimit = 2.0f;
+	WaitForStationaryActorStepDefinition.NativeTickEvent.BindLambda([this](float DeltaTime) {
+		UGameInstance* GameInstance = GetGameInstance();
+		const FString WorkerId = GameInstance->GetSpatialWorkerId();
+		for (TActorIterator<ATestMovementCharacter> Iter(GetWorld()); Iter; ++Iter)
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp:57]: As Simon told me about in a different PR, you can use a `TActorRange` here

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp [hunk 11]
```diff
@@ -76,89 +86,84 @@ void ASpatialTestCharacterMigration::PrepareTest()
...
+				Character->AddMovementInput(FVector(1.0f, 0.0f, 0.0f), 10.0f, true);
+				float PeakSpeed = Character->GetPeakSpeedInWindow();
+				AssertValue_Float(PeakSpeed, EComparisonMethod::Less_Than, 650.0f, TEXT("Actor not speeding"));
+			}
 
-		if (bCharacterReachedDestination)
-		{
-			AssertTrue(bCharacterReachedDestination, TEXT("Player character has reached the destination on the autonomous proxy."));
-			FinishStep();
+			bool bReachDestination = GetTargetDistanceOnLine(Origin, Destination, Character->GetActorLocation()) > -40.0f; // 20cm overlap
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp:107]:
```suggestion
			bool bReachDestination = GetTargetDistanceOnLine(Origin, Destination, Character->GetActorLocation()) > -40.0f; // 40cm overlap
```
? But also not actually sure how this is measured

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp [hunk 12]
```diff
@@ -76,89 +86,84 @@ void ASpatialTestCharacterMigration::PrepareTest()
...
+				Character->AddMovementInput(FVector(-1.0f, 0.0f, 0.0f), 10.0f, true);
+				float PeakSpeed = Character->GetPeakSpeedInWindow();
+				AssertValue_Float(PeakSpeed, EComparisonMethod::Less_Than, 650.0f, TEXT("Actor not speeding"));
+			}
 
-	// Universal setup step to create the TriggerBox and to set the helper variable
-	AddStep(TEXT("UniversalSetupStep"), FWorkerDefinition::AllWorkers, nullptr, [this]() {
-		bCharacterReachedDestination = false;
-		bCharacterReachedOrigin = false;
+			bool bReachDestination = GetTargetDistanceOnLine(Destination, Origin, Character->GetActorLocation()) > -40.0f; // 20cm overlap
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp:140]:
```suggestion
			bool bReachDestination = GetTargetDistanceOnLine(Destination, Origin, Character->GetActorLocation()) > -40.0f; // 40cm overlap
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/TestActors/TestMovementCharacter.cpp [hunk 13]
```diff
@@ -25,8 +28,8 @@ ATestMovementCharacter::ATestMovementCharacter()
 	SphereComponent->SetVisibility(true);
 	SphereComponent->SetupAttachment(GetCapsuleComponent());
 
-	FVector CameraLocation = FVector(300.0f, 0.0f, 75.0f);
-	FRotator CameraRotation = FRotator::MakeFromEuler(FVector(0.0f, -10.0f, 180.0f));
+	FVector CameraLocation = FVector(300.0f, 300.0f, 75.0f);
+	FRotator CameraRotation = FRotator::MakeFromEuler(FVector(0.0f, -10.0f, 225.0f));
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/TestActors/TestMovementCharacter.cpp:32]: did we have it wrong?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/TestActors/TestMovementCharacter.cpp (hunk 5 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/TestActors/TestMovementCharacter.cpp:64]: I would normally do `SpeedWindowIndex = (SpeedWindowIndex + 1) % SpeedWindowSize;`, but I know some people have expressed unreserved disgust at such a method, so leaving it here as an option.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/TestActors/TestMovementCharacter.cpp (hunk 13 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/TestActors/TestMovementCharacter.cpp:32]: I think the new spawning just rotated the actors slightly... I just adjusted the camera accordingly.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Private/TestMaps/Spatial2WorkerFullInterestMigrationMap.cpp (hunk 1 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/Private/TestMaps/Spatial2WorkerFullInterestMigrationMap.cpp:29]: Have opted to add them into one cpp. Deriving from a common class has the issue of the common class being "generated" when generate maps. Perhaps there is a nice way around this?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp (hunk 2 above)
Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp:42]: Think GetWorkerEntityPosition does not actually tell you where the boundary is? With this 2 zone grid, I get a "WorkerEntityPosition " of { 250000, 0, 0 }.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/TestActors/TestMovementCharacter.cpp (hunk 13 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/TestActors/TestMovementCharacter.cpp:32]: Ah, okay, just that this is used by other tests as well (unfortunately). So not sure what the effects elsewhere are.
It's a sphere anyway, so kind of hard to judge.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp (hunk 10 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp:57]: Ah, you can do:
```
for (ATestMovementCharacter* Character : TActorRange<ATestMovementCharacter>(GetWorld())) {}
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Private/TestMaps/Spatial2WorkerFullInterestMigrationMap.cpp (hunk 1 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/Private/TestMaps/Spatial2WorkerFullInterestMigrationMap.cpp:29]: Ah, true - you can theoretically override `ShouldGenerateMap()` to control this, but it's probably more hassle than it's worth.
I think having it in the same file, possibly using a file-local function should be enough.

Comment (MatthewSandfordImprobable) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/Private/TestMaps/Spatial2WorkerFullInterestMigrationMap.cpp:29]: You unfortunately AddActorToLevel is protected so I can't pull out a function.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp [hunk 14]
```diff
@@ -75,90 +84,117 @@ void ASpatialTestCharacterMigration::PrepareTest()
...
+			if (bLoadBalanceStrategyValid)
+			{
+				const VirtualWorkerId ExpectedVirtualWorker = 2;
+				const VirtualWorkerId LocalVirtualWorker = SpatialNetDriver->LoadBalanceStrategy->GetLocalVirtualWorkerId();
+				bShouldRunRequires = LocalVirtualWorker == ExpectedVirtualWorker;
+
+				if (bShouldRunRequires)
+				{
+					const VirtualWorkerId VirtualWorker = SpatialNetDriver->LoadBalanceStrategy->WhoShouldHaveAuthority(*Character);
+					RequireEqual_Int(VirtualWorker, ExpectedVirtualWorker, FString::Printf(TEXT("%s on %s crossed boundary"), *Character->GetName(), *WorkerId));
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp:116]: Can't you run this on both servers? If you're running with no interest, you won't have a ATestMovementCharacter actor to operate on. And if you're running with interest, the call to `WhoShouldHaveAuthority` should still return the same value.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp (hunk 2 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp:43]: nit: might be better to rename these - PositionOnServerOne etc.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/Private/TestMaps/Spatial2WorkerFullInterestMigrationMap.cpp (hunk 1 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/Private/TestMaps/Spatial2WorkerFullInterestMigrationMap.cpp]: whoops

**Review by m-samiec — APPROVED** [review]

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp [hunk 15]
```diff
@@ -38,19 +39,27 @@ ASpatialTestCharacterMigration::ASpatialTestCharacterMigration()
...
-	FSpatialFunctionalTestStepDefinition ResetStepDefinition(/*bIsNativeDefinition*/ true);
-	ResetStepDefinition.StepName = TEXT("Reset");
-	ResetStepDefinition.TimeLimit = 0.0f;
-	ResetStepDefinition.NativeStartEvent.BindLambda([this]() {
-		bCharacterReachedDestination = false;
-		bCharacterReachedOrigin = false;
+	// Wait for actor to be stationary
+	FSpatialFunctionalTestStepDefinition WaitForStationaryActorStepDefinition(/*bIsNativeDefinition*/ true);
+	WaitForStationaryActorStepDefinition.StepName = TEXT("WaitForStationaryActorStepDefinition");
+	WaitForStationaryActorStepDefinition.TimeLimit = 2.0f;
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp:54]: Why deviate from the default? E.g. why not just omit this line?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp [hunk 16]
```diff
@@ -38,19 +39,27 @@ ASpatialTestCharacterMigration::ASpatialTestCharacterMigration()
...
+	WaitForStationaryActorStepDefinition.NativeTickEvent.BindLambda([this](float DeltaTime) {
+		UGameInstance* GameInstance = GetGameInstance();
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp:56]: const?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp [hunk 17]
```diff
@@ -75,90 +84,123 @@ void ASpatialTestCharacterMigration::PrepareTest()
 
 	// Move character forward
 	FSpatialFunctionalTestStepDefinition MoveForwardStepDefinition(/*bIsNativeDefinition*/ true);
-	MoveForwardStepDefinition.StepName = TEXT("Client1MoveForward");
-	MoveForwardStepDefinition.TimeLimit = 0.0f;
+	MoveForwardStepDefinition.StepName = TEXT("MoveForwardStepDefinition");
+	MoveForwardStepDefinition.TimeLimit = 5.0f;
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp:88]: Why deviate from the default?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMovement/CharacterMovementTestGameMode.cpp [hunk 18]
```diff
@@ -1,10 +1,28 @@
...
 ACharacterMovementTestGameMode::ACharacterMovementTestGameMode(const FObjectInitializer& ObjectInitializer)
 	: Super(ObjectInitializer)
 {
 	DefaultPawnClass = ATestMovementCharacter::StaticClass();
 }
+
+void ACharacterMovementTestGameMode::BeginPlay()
+{
+	Super::BeginPlay();
+	for (TActorIterator<APlayerStart> Iter(GetWorld()); Iter; ++Iter)
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMovement/CharacterMovementTestGameMode.cpp:18]: TActorRange?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/TestActors/TestMovementCharacter.cpp [hunk 19]
```diff
@@ -41,8 +42,23 @@ ATestMovementCharacter::ATestMovementCharacter()
 	CameraComponent->SetRelativeRotation(CameraRotation);
 #endif
 	CameraComponent->SetupAttachment(GetCapsuleComponent());
+
+	UCharacterMovementComponent* CharacterMovementComponent = GetCharacterMovement();
+	CharacterMovementComponent->MaxSimulationIterations = 14;
```
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/TestActors/TestMovementCharacter.cpp:47]: comment to explain this 14?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp [hunk 20]
```diff
@@ -38,19 +39,25 @@ ASpatialTestCharacterMigration::ASpatialTestCharacterMigration()
...
-	ResetStepDefinition.TimeLimit = 0.0f;
-	ResetStepDefinition.NativeStartEvent.BindLambda([this]() {
-		bCharacterReachedDestination = false;
-		bCharacterReachedOrigin = false;
+	// Wait for actor to be stationary
+	FSpatialFunctionalTestStepDefinition WaitForStationaryActorStepDefinition(/*bIsNativeDefinition*/ true);
+	WaitForStationaryActorStepDefinition.StepName = TEXT("WaitForStationaryActorStepDefinition");
+	WaitForStationaryActorStepDefinition.NativeTickEvent.BindLambda([this](float DeltaTime) {
+		const UGameInstance* GameInstance = GetGameInstance();
+		const FString WorkerId = GameInstance->GetSpatialWorkerId();
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestCharacterMigration/SpatialTestCharacterMigration.cpp:56]: nit, could also just do `GetGameInstance()->GetSpatialWorkerId()` if we don't expect the game instance to be null, but nbd

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/TestActors/TestMovementCharacter.cpp [hunk 21]
```diff
@@ -41,8 +42,24 @@ ATestMovementCharacter::ATestMovementCharacter()
 	CameraComponent->SetRelativeRotation(CameraRotation);
 #endif
 	CameraComponent->SetupAttachment(GetCapsuleComponent());
+
+	// Increase MaxSimulationIterations to 14 so that we do not get warnings inside UCharacterMovementComponent::GetSimulationTimeStep
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/TestActors/TestMovementCharacter.cpp:46]: ah, funny, I have seen this warning in EP, maybe we should take a look at some point

**Review by mironec — APPROVED** [review]
Seems good

**Comment by mironec** [comment]
hm, seems like the test is failing in CI, may need more debugging + maybe running multiple times
